crates/router/src/core/passive_churn_recovery/types.rs

```
@@ -0,0 +1,309 @@
+                let psync_data = core_pcr::call_psync_api(state, payment_id, pcr_data)
+                    .await
+                    .change_context(errors::RecoveryError::PaymentCallFailed)
+                    .attach_printable("Error while executing the Psync call")?;
+                let payment_attempt = psync_data
+                    .payment_attempt
+                    .get_required_value("Payment Attempt")
+                    .change_context(errors::RecoveryError::ValueNotFound)
+                    .attach_printable("Error while executing the Psync call")?;
+                Self::PsyncTask(payment_attempt)
+            }
+            _ => Self::InvalidTask,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub enum Action {
+    SyncPayment,
+    RetryPayment,
+    TerminalFailure,
+    SuccessfulPayment,
+    ReviewPayment,
+    ManualReviewAction,
+}
+impl Action {
+    pub async fn execute_payment(
+        db: &dyn StorageInterface,
+        merchant_id: &id_type::MerchantId,
+        payment_intent: &PaymentIntent,
+        process: &storage::ProcessTracker,
+    ) -> RecoveryResult<Self> {
+        // call the proxy api
+        let response = call_proxy_api::<api_types::Authorize>(payment_intent);
+        // handle proxy api's response
+        match response {
+            Ok(payment_data) => match payment_data.payment_attempt.status.foreign_into() {
+                PCRAttemptStatus::Succeeded => Ok(Self::SuccessfulPayment),
+                PCRAttemptStatus::Failed => {
+                    Self::decide_retry_failure_action(db, merchant_id, process.clone()).await
+                }
+
+                PCRAttemptStatus::Processing => Ok(Self::SyncPayment),
+                PCRAttemptStatus::InvalidAction(action) => {
+                    logger::info!(?action, "Invalid Payment Status For PCR Payment");
+                    Ok(Self::ManualReviewAction)
+                }
+            },
+            Err(_) =>
```

Avoid discarding the error, log the error